src/main/java/org/apache/spark/shuffle/ucx/reducer/UcxShuffleClient.java

```diff
@@ -0,0 +1,140 @@
+  private void submitFetchOffsets(UcpEndpoint endpoint, ShuffleBlockId[] blockIds,
+                                  long[] dataAddresses, RegisteredMemory offsetMemory) {
+    ByteBuffer driverMetadata =  workerWrapper.fetchDriverMetadataBuffer(handle.shuffleId()).data();
+    for (int i = 0; i < blockIds.length; i++) {
+      ShuffleBlockId blockId = blockIds[i];
+      // Get block offset
+      int mapIdBlock = blockId.mapId() *
+        (int) ucxShuffleManager.ucxShuffleConf().metadataBlockSize();
+      int offsetWithinBlock = 0;
+      // Parse metadata fetched from driver.
+      long offsetAdress = driverMetadata.getLong(mapIdBlock + offsetWithinBlock);
+      offsetWithinBlock += 8;
+      long dataAddress = driverMetadata.getLong(mapIdBlock + offsetWithinBlock);
+      dataAddresses[i] = dataAddress;
+      offsetWithinBlock += 8;
+
+      // Unpack Rkeys for this workerWrapper
+      if (offsetRkeysCache.get(blockId.mapId()) == null ||
+        dataRkeysCache.get(blockId.mapId()) == null) {
+        int offsetRKeySize = driverMetadata.getInt(mapIdBlock + offsetWithinBlock);
+        offsetWithinBlock += 4;
+        int dataRkeySize = driverMetadata.getInt(mapIdBlock + offsetWithinBlock);
+        offsetWithinBlock += 4;
+
+        if (offsetRKeySize <= 0 || dataRkeySize <= 0) {
+          throw new UcxException("Wrong rkey size.");
+        }
+        final ByteBuffer rkeyCopy = driverMetadata.slice();
+        rkeyCopy.position(mapIdBlock + offsetWithinBlock)
+          .limit(mapIdBlock + offsetWithinBlock + offsetRKeySize);
+
+        offsetWithinBlock += offsetRKeySize;
+
+        offsetRkeysCache.computeIfAbsent(blockId.mapId(), mapId -> endpoint.unpackRemoteKey(rkeyCopy));
+
+        rkeyCopy.position(mapIdBlock + offsetWithinBlock)
+          .limit(mapIdBlock + offsetWithinBlock + dataRkeySize);
+
+        dataRkeysCache.computeIfAbsent(blockId.mapId(), mapId -> endpoint.unpackRemoteKey(rkeyCopy));
+      }
+
+      endpoint.getNonBlockingImplicit(
+        offsetAdress + blockId.reduceId() * 8L, offsetRkeysCache.get(blockId.mapId()),
+        UcxUtils.getAddress(offsetMemory.getBuffer()) + (i * 16), 16);
+    }
+  }
+
+  /**
+   * Reducer entry point. Fetches remote blocks, using 2 ucp_get calls.
+   * This method is inside ShuffleFetchIterator's for loop over hosts.
+   * First fetches block offset from index file, and then fetches block itself.
+   */
+  @Override
+  public void fetchBlocks(String host, int port, String execId,
+                          String[] blockIds, BlockFetchingListener listener,
+                          DownloadFileManager downloadFileManager) {
+    long startTime = System.currentTimeMillis();
+
+    BlockManagerId blockManagerId = BlockManagerId.apply(execId, host, port, Option.empty());
+    UcpEndpoint endpoint = workerWrapper.getConnection(blockManagerId);
+
+    long[] dataAddresses = new long[blockIds.length];
+
+    RegisteredMemory offsetMemory = mempool.get(16 * blockIds.length);
```

Comment:
create constant for '16' to be used by both mapper and reducer (on mapper it could be an assertion that '16' * num_blocks == mapped file size)